torch/testing/_internal/common_utils.py

```diff
@@ -5312,3 +5312,36 @@ def repl_frame(m):
         s = re.sub(r"Cannot export model.+\n\n", "", s)
     s = re.sub(r" +$", "", s, flags=re.MULTILINE)
     return s
+
+def recover_orig_fp32_precision():
+    @contextlib.contextmanager
+    def recover():
+        old_mkldnn_conv_p = torch.backends.mkldnn.conv.fp32_precision
+        old_mkldnn_rnn_p = torch.backends.mkldnn.rnn.fp32_precision
+        old_mkldnn_matmul_p = torch.backends.mkldnn.matmul.fp32_precision
+        old_cudnn_conv_p = torch.backends.cudnn.conv.fp32_precision
+        old_cudnn_rnn_p = torch.backends.cudnn.rnn.fp32_precision
+        old_cuda_matmul_p = torch.backends.cuda.matmul.fp32_precision
+        old_matmul_precision = torch.get_float32_matmul_precision()
+        old_cubls_tf32 = torch.backends.cuda.matmul.allow_tf32
+        try:
+            yield
+        finally:
+            torch.backends.mkldnn.conv.fp32_precision = old_mkldnn_conv_p
+            torch.backends.mkldnn.rnn.fp32_precision = old_mkldnn_rnn_p
+            torch.backends.mkldnn.matmul.fp32_precision = old_mkldnn_matmul_p
+            torch.backends.cudnn.conv.fp32_precision = old_cudnn_conv_p
+            torch.backends.cudnn.rnn.fp32_precision = old_cudnn_rnn_p
+            torch.backends.cuda.matmul.fp32_precision = old_cuda_matmul_p
+            torch.set_float32_matmul_precision(old_matmul_precision)
+            torch.backends.cuda.matmul.allow_tf32 = old_cubls_tf32
+
+    def wrapper(f):
+        @functools.wraps(f)
+        def wrapped(*args, **kwargs):
+            with recover():
+                f(*args, **kwargs)
```

Comment:
Do we really need 4 layers of nesting here?

Reply:
Thanks for the reminder here.
This should be a 3 layers nesting for wrap fn and 1 layer for contextmanager.
I took a look for some existing [decorators](https://github.com/pytorch/pytorch/blob/a7a53b796b1fcde9a52c86d29d88e12bbf2b4092/torch/testing/_internal/common_utils.py#L1613-L1629) and I think 3 layers nesting to wrap an fn with `args, kwargs` is needed. And the contextmanager is also a good way if we wish to have pre-post process for this fn. Do you have more detailed suggestions here?

Reply:
You can make the decorator applies as @recover_orig_fp32_precision and not @recover_orig_fp32_precision()
to remove one layer. Making a new recover each time.

Also the recover function can be removed and you can inline the try/except block directly.

Reply:
Thanks for advice here! I removed 1 layer by using `@recover_orig_fp32_precision`.
As for inline `try/except`, I fill it's better to keep `def recover()` here I feel this is a common usage like other places if we want to stash and recover something before/after a function.